src/monodroid/jni/monodroid-glue.c

```diff
@@ -1278,91 +1278,237 @@ gc_is_bridge_object (MonoObject *object)
-					mono.mono_class_get_name (mono.mono_object_get_class (reffed_obj)));
-#endif
+
 		return 1;
 	}
 
 	(*env)->ExceptionClear (env);
-#if DEBUG
-	if (gc_spew_enabled)
-		log_error (LOG_GC, "Missing monodroidAddReference method for object of class %s.%s",
-				mono.mono_class_get_namespace (klass),
-				mono.mono_class_get_name (klass));
-#endif
 	(*env)->DeleteLocalRef (env, java_class);
 	return 0;
 }
 
+// add_reference can work with objects which are either MonoObjects with java peers, or raw jobjects
+typedef struct {
+	mono_bool is_mono_object;
+	union {
+		MonoObject *obj;
+		jobject jobj;
+	};
+} AddReferenceTarget;
+
+// Given a target, extract the bridge_info (if a mono object) and handle. Return success.
+static mono_bool
+load_reference_target (AddReferenceTarget target, MonoJavaGCBridgeInfo** bridge_info, jobject *handle)
+{
+	if (target.is_mono_object) {
+		*bridge_info = get_gc_bridge_info_for_object (target.obj);
+		if (!*bridge_info)
+			return FALSE;
+		mono.mono_field_get_value (target.obj, (*bridge_info)->handle, handle);
+	} else {
+		*handle = target.jobj;
+	}
+	return TRUE;
+}
+
+#if DEBUG
+// Allocate and return a string describing a target
+static char *describe_target (AddReferenceTarget target) {
+	if (target.is_mono_object) {
+		MonoClass *klass = mono.mono_object_get_class (target.obj);
+		return monodroid_strdup_printf ("object of class %s.%s",
+			mono.mono_class_get_namespace (klass),
+			mono.mono_class_get_name (klass));
+	}
+	else
+		return monodroid_strdup_printf ("<temporary object %p>", target.jobj);
+}
+#endif
+
+// Add a reference from one target to another. If the "from" target is a mono_object, it must be a user peer
+static mono_bool
+add_reference (JNIEnv *env, AddReferenceTarget target, AddReferenceTarget reffed_target)
+{
+	MonoJavaGCBridgeInfo *bridge_info = NULL, *reffed_bridge_info = NULL;
+	jobject handle, reffed_handle;
+
+	if (!load_reference_target (target, &bridge_info, &handle))
+		return FALSE;
+
+	if (!load_reference_target (reffed_target, &reffed_bridge_info, &reffed_handle))
+		return FALSE;
+
+	mono_bool success = add_reference_jobject (env, handle, reffed_handle);
+
+	// Flag MonoObjects so they can be cleared in gc_cleanup_after_java_collection.
+	// Java temporaries do not need this because the entire GCUserPeer is discarded.
+	if (success && target.is_mono_object) {
+		int ref_val = 1;
+		mono.mono_field_set_value (target.obj, bridge_info->refs_added, &ref_val);
+	}
+
+#if DEBUG
+	if (gc_spew_enabled) {
+		char *description = describe_target (target),
+			 *reffed_description = describe_target (reffed_target);
+
+		if (success)
+			log_warn (LOG_GC, "Added reference for %s to %s", description, reffed_description);
+		else
+			log_error (LOG_GC, "Missing monodroidAddReference method for %s", description);
+
+		free (description);
+		free (reffed_description);
+	}
+#endif
+
+	return success;
+}
+
+// Create a target
+AddReferenceTarget mono_object_target (MonoObject *obj) {
```

Comment:
the mono_ is because this is a MonoObject. there is no such thing as a monodroid_object. maybe i could prefix both _target methods with make_ so it is clear the mono_ is not a namespace ?